README.md
"# Bloons-DQL" 


Paint Tool Sai 2.0 x64 (2019)/history.txt
--------------------------------------------------------------------------------------------------------------------------------------------
2019-08-12

[Fixed Bugs]

- [Symmetric Ruler] Cnannot convert symmetric shapes to regular shapes correctly by [Convert Symmetric Shapes to Regular Shapes].

- [Lasso Tool] Cannot create a traiangle shape in the polygonal mode.

- [PSD File] Cannot open a PSD file including folder structures deeper than 5 levels.
  (There is a possibility that this fix does not work because this problem was fixed under an unknown PSD specification.)


--------------------------------------------------------------------------------------------------------------------------------------------
2019-08-07

[Fixed Bugs]

- [Marker] (32bit ver. only) Does not work correctly.

- [Symmetric Ruler] The angle of point symmetric drawing is not correct.

- [Symmetric Ruler] The affection of [Scattering] parameter of scattering is not symmetry.

- [Symmetric Ruler] Access violations occur when uses Smudge tool with a brush texture.

- [Perspective Ruler] The detection button of the red vanishing point is disabled for 1VP and 2VP perspective ruler.

- [Linework Layer] Strokes are not rendered correctly.


[Changes]

- [Symmetric Ruler] Changed to render shapes of bristle and scattering as line symmetry.

- [Linework Layer] Improved rendering speed for multiple strokes operation.


--------------------------------------------------------------------------------------------------------------------------------------------
2019-07-29

[Fixed Bugs]

- [Scratch Pad] The program does not start with an error if the saved file of scratch-pad is not existing.

- [Linework Layer] Often create incorrect srokes. (An stroke error will occur when opened *.sai2 file that included these strokes).

- [Symmetric Ruler] Cannot draw correctly if the ellipse ruler is turning on.

- [Symmetric Ruler] The shape directions are not symmetric.

- [Marquee Tool / Lasso Tool] The intersection operation(Alt + Shift) does not work correctly.


--------------------------------------------------------------------------------------------------------------------------------------------
2019-07-28

 * Caution! : Recovery data created by this version is not compatible with older versions.

[New Features]

- Symmetric Ruler
  + This ruler can be created from the menu of perspective ruler on the layer panel.

  + Supported line symmmetry and point symmetry drawing. The maximum number of division is 20.

  + This ruler is applicable to drawing on a linework layer, but each mirrored stroke does not have a substance.
    [Convert Symmetric Shapes to Regular Shapes] in [Layer] menu converts mirrored strokes to regular strokes.


[Fixed Bugs]

- [Scattering Tool] Cannot paint to the scratch pad.

- [Scattering Tool] An eraser mark is not added to the mouse cursor even if the current color is transparent.

- [Brush Tools] [Preserve Pixel Opacity] in an ancestor folder is not applied to brush painting.

- [Binary Pen] The size of brush circle react to the pen pressure even if [Algorithm] is [Simple].

- [Eraser Tool(for Linework)] The program sometimes freezes when erase some particular stroke.

- [Linework Layer] The rendering result of a layer paper is different from older version.

- [Shape Layers] The hitting range of a control point is large than Ver.1.

- [Options Dialog] The program fails to start if specified an invalid path for the path of recovery data.


[Changes]

- [Layer] Show Alone mode is changed to chase a current layer among only sibling layers.

- [Brush Tools] Tuned the quality of small blotmap brush(0～6px).

- [Brush Tools] Added again [Angle] and [Angle Jitter] to the scattering form of the regular brushes.

- [Scattering Tools] Changed to apply the gaussian method to the shape size.
  (Render near shapes large and render far shapes small)

- [Scattering] Changed to apply the horizontal direction of view to [Angle] if [Angle Control] is [Pen Direction].

- [Eraser Tool(for Linework)] Added the option of multiple linework layer erasing.

- [Lasso Tool] Added [Polygonal + Freeform] mode.

- [Marquee Tool / Lasso Tool] Added the mode to intersect a selection (Alt + Shift)

- [Perspective Rulers] Changed to preserve the settings of vanishing point detection into the each ruler.

- [Perspective Grid] Changed to preserve the settings of grid snap and vanishing point detection into the each ruler.

- [Change Canvas Resolution] Changed the behavior of [Printing Resolution] to affect [Width] and [Size]
  when the size unit is "%" or "pixels" too.

- [Pen Tablet Control] Changed to suppress WinTab errors of malfunctioning driver.


And some other small changes and some small bug fixes...


--------------------------------------------------------------------------------------------------------------------------------------------
2019-05-21

[Fixed Bugs]

- [Scattering Tool] Shapes that the ordinal is greater than the number of brush forms of other tools do not work.


[Changes]

- [Pen Tablet Control] Changed to get not the button mapping of pen other if the mode is Ver.1 compatible.
  (To avoid a bug of Wacom driver that WinTab errors occur when activated the program window)


--------------------------------------------------------------------------------------------------------------------------------------------
2019-05-19

[Fixed Bugs]

- [Brush Tools] [HQ Sampling] of scattering form does not work

- [Brush Tools] An access violation occurs if [HQ Sampling] of scattering form is OFF and the brush size is small.

- [Brush Tools] The names of scaling quality of scattering from is not [HQ Down Sampling].

- [Marker] The brush form does not work.

- [Scattering Tool] The selection of [Angle Control] dropdown list is not correct.

- [GUI Control] Access violations occur when shape layers was deleted by Undo of Redo.



code1.py
import pygame as pg
import pygame.mouse as mouse
import numpy as np
import random
import sys
import math
from pygame import gfxdraw

def draw():
	surface.blit(background_sprite, (0, 0))
	for i in range(rows):
		for j in range(columns):
			if bloons_data[i][j] > 0:
				surface.blit(bloon_sprites[bloons_data[i][j] - 1], (j * cell_width, i * cell_height))
	surface.blit(monkey_sprite, monkey_sprite_position)
	surface.blit(arm_sprite, arm_sprite_position)
	surface.blit(dart_sprite, dart_sprite_position)
	pg.display.update()

#	Variables
framerate = 120

rows = 10
columns = 20

cell_width = 57
cell_height = 80

mouse_angle = 0

	#	Bloons
bloons_data = []
for i in range(rows):
	bloons_data.append([])
	for j in range(columns):
		if i > 0 and i <= 6 and j > 7 and j <= 17:
			bloons_data[i].append(1 + random.randrange(4))
		else:
			bloons_data[i].append(0)

	#	Monkey
monkey_initial_position = [2, 7]	# x and y positions (on the grid)

	#	Dart
flying = False
dart_speed = 1920
dart_velocity = [0, 0]
dart_gravity = 3840

#	Pygame
pg.init()
surface = pg.display.set_mode((columns * cell_width, rows * cell_height))
clock = pg.time.Clock()

#	Sprites
background_sprite = pg.image.load("sprites/background.png")

bloon_sprites = []
bloon_sprites.append(pg.image.load("sprites/red_bloon.png"))
bloon_sprites.append(pg.image.load("sprites/green_bloon.png"))
bloon_sprites.append(pg.image.load("sprites/blue_bloon.png"))
bloon_sprites.append(pg.image.load("sprites/yellow_bloon.png"))

for i in range(4):
	bloon_sprites[i] = pg.transform.scale(bloon_sprites[i], (cell_width, cell_height))
	#bloon_sprites[i] = pg.transform.rotate(bloon_sprites[i], random.randrange(11) - 5)

monkey_sprite = pg.image.load("sprites/monkey_centered.png")
monkey_sprite = pg.transform.scale(monkey_sprite, (196, 196))
monkey_sprite_position = [monkey_initial_position[0] * cell_width - 98, monkey_initial_position[1] * cell_height - 98]

arm_angle = 0
monkey_arm = pg.image.load("sprites/monkey_arm_centered.png")
arm_sprite = pg.transform.scale(monkey_arm, (196, 196))
arm_sprite = pg.transform.rotate(arm_sprite, arm_angle)
arm_sprite_position = [0, 0]

dart = pg.image.load("sprites/dart.png")
dart_sprite = pg.transform.scale(dart, (136, 136))
dart_sprite = pg.transform.rotate(dart_sprite, mouse_angle)
dart_sprite_position = [0, 0]
dart_real_position = [0, 0]

#	Juego
run = True
while run:
	for event in pg.event.get():
		if event.type == pg.QUIT:
			run = False

	#keys = pg.key.get_pressed()
	[mouse_x, mouse_y] = mouse.get_pos()

	#left_onpress = not left_click
	left_click = mouse.get_pressed(3)[0]
	if left_click and not flying:
		flying = True
		dart_velocity[0] = dart_speed * np.cos(mouse_angle * np.pi / 180.0)
		dart_velocity[1] = - dart_speed * np.sin(mouse_angle * np.pi / 180.0)
		magnitude = np.sqrt(dart_velocity[0]**2 + dart_velocity[1]**2)
		dart_real_position[0] = monkey_sprite_position[0] + 98 + cell_width / 2 + dart_velocity[0] / magnitude
		dart_real_position[1] = monkey_sprite_position[1] + 98 + cell_height / 2 + dart_velocity[1] / magnitude
	#left_onpress = left_onpress and left_click
	right_click = mouse.get_pressed(3)[2]
	if right_click:
		flying = False

	#if (left_click or right_click) and search_state[0] == 2:
	#	search_state[0] = 0
	#	state = "Idle"
	#	state_color[0] = white
	#	state_text[0] = font.render(state, True, state_color[0], None)

	#check_button_hovers(mouse_x, mouse_y, left_click, left_onpress)

	draw()
	delta_time = clock.tick_busy_loop(60) / 1000.0
	delta_time = 0.0008
	pg.display.set_caption("Bloons AI - FPS: " + str(int(clock.get_fps())))

	mouse_angle = - math.atan2(mouse_y - (monkey_sprite_position[1] + 98 + 33), mouse_x - (monkey_sprite_position[0] + 98 + 5)) * 180.0 / np.pi

	arm_angle = mouse_angle + 135
	arm_sprite = pg.transform.scale(monkey_arm, (196, 196))
	arm_sprite = pg.transform.rotate(arm_sprite, arm_angle)

	dx = 98 * (np.cos(arm_angle * np.pi / 180.0) + np.sin(arm_angle * np.pi / 180.0))
	dy = 98 * (np.cos(arm_angle * np.pi / 180.0) - np.sin(arm_angle * np.pi / 180.0))
	offset = max(abs(dx), abs(dy)) - 98

	arm_sprite_position[0] = monkey_sprite_position[0] - offset + 5
	arm_sprite_position[1] = monkey_sprite_position[1] - offset + 33

	if not flying:
		arm_length = 30
		dart_sprite = pg.transform.scale(dart, (136, 136))
		dart_sprite = pg.transform.rotate(dart_sprite, mouse_angle)

		dx = 68 * (np.cos(mouse_angle * np.pi / 180.0) + np.sin(mouse_angle * np.pi / 180.0))
		dy = 68 * (np.cos(mouse_angle * np.pi / 180.0) - np.sin(mouse_angle * np.pi / 180.0))
		offset = max(abs(dx), abs(dy)) - 68

		dart_sprite_position[0] = monkey_sprite_position[0] + 98 + 5 + arm_length * np.cos(arm_angle * np.pi / 180.0) - 68 - offset
		dart_sprite_position[1] = monkey_sprite_position[1] + 98 + 33 - arm_length * np.sin(arm_angle * np.pi / 180.0) - 68 - offset
	else:
		dart_velocity[1] += dart_gravity * delta_time
		dart_angle = - math.atan2(dart_velocity[1], dart_velocity[0]) * 180 / np.pi
		dart_real_position[0] += dart_velocity[0] * delta_time
		dart_real_position[1] += dart_velocity[1] * delta_time

		dart_sprite_position[0] = dart_real_position[0] - 68
		dart_sprite_position[1] = dart_real_position[1] - 68
		dart_sprite = pg.transform.scale(dart, (136, 136))
		dart_sprite = pg.transform.rotate(dart_sprite, dart_angle)

		dx = 68 * (np.cos(mouse_angle * np.pi / 180.0) + np.sin(mouse_angle * np.pi / 180.0))
		dy = 68 * (np.cos(mouse_angle * np.pi / 180.0) - np.sin(mouse_angle * np.pi / 180.0))
		offset = max(abs(dx), abs(dy)) - 68

		dart_sprite_position[0] += - offset
		dart_sprite_position[1] += - offset

pg.quit()

code.py
import pygame as pg
import pygame.mouse as mouse
import numpy as np
import random
import sys
import math
from pygame import gfxdraw

sign_fix = -1

def draw():
	background_sprite.draw()
	for i in range(columns):
		for j in range(rows):
			if bloon_sprites[i][j] != None:
				bloon_sprites[i][j].draw()
	monkey_sprite.draw()
	arm_sprite.draw()
	dart_sprite.draw()
	pg.display.update()

class Sprite:
	def __init__(self, x, y, img_src, w, h):
		self.x = x
		self.y = y
		self.w = w
		self.h = h
		self.tx = x - w / 2
		self.ty = y - h / 2
		self.tw = w
		self.th = h
		self.source = pg.image.load("sprites/" + img_src)
		self.image = pg.transform.scale(self.source, (w, h))
		self.angle = 0

	def set_center(self, x, y):
		self.x = x
		self.y = y
		self.tx = x - self.w / 2
		self.ty = y - self.h / 2

	def update(self):
		self.tx = self.x - self.tw / 2
		self.ty = self.y - self.th / 2

	def set_angle(self, new_angle):
		if self.angle == new_angle:
			return
		self.angle = new_angle
		rad_angle = self.angle * np.pi / 180.0
		self.image = pg.transform.scale(self.source, (self.w, self.h))
		self.image = pg.transform.rotate(self.image, sign_fix * self.angle)
		if np.cos(rad_angle) * np.sin(rad_angle) >= 0:
			self.tw = np.absolute(self.w * np.cos(rad_angle) + self.h * np.sin(rad_angle))
			self.th = np.absolute(self.w * np.sin(rad_angle) + self.h * np.cos(rad_angle))
		else:
			self.tw = np.absolute(self.w * np.cos(rad_angle) - self.h * np.sin(rad_angle))
			self.th = np.absolute(self.w * np.sin(rad_angle) - self.h * np.cos(rad_angle))
		self.tx = self.x - self.tw / 2
		self.ty = self.y - self.th / 2

	def draw(self):
		surface.blit(self.image, (self.tx, self.ty))

#	Variables
framerate = 60
delta_time = 0.0008

columns = 20
rows = 10

cell_width = 57
cell_height = 80

width = columns * cell_width
height = rows * cell_height

mouse_angle = 0

	#	Bloons
bloons_data = []
for i in range(columns):
	bloons_data.append([])
	for j in range(rows):
		if i >= columns / 2 and i < columns - 1 and (j == 2 or j == 4 or j == 6):
			bloons_data[i].append(1)
		else:
			bloons_data[i].append(0)

	#	Monkey
monkey_initial_position = [2, 7]	# x and y positions (on the grid)

	#	Dart
flying = False
dart_speed = 2400
dart_velocity = [0, 0]
dart_gravity = 3840

#	Pygame
pg.init()
pg.display.set_icon(pg.image.load("sprites/monkey.png"))
surface = pg.display.set_mode((columns * cell_width, rows * cell_height))
clock = pg.time.Clock()

#	Sprites
background_sprite = Sprite(width / 2, height / 2, "background.png", width, height)
bloon_img = "red_bloon", "green_bloon", "blue_bloon", "yellow_bloon"

bloon_sprites = []
for i in range(columns):
	bloon_sprites.append([])
	for j in range(rows):
		if bloons_data[i][j] == 1:
			bloon_sprites[i].append(Sprite((i + 0.5) * cell_width, (j + 0.5) * cell_height, bloon_img[random.randrange(4)] + ".png", cell_width, cell_height))
			#bloon_sprites[i][j].set_angle(random.randrange(11) - 5)
		else:
			bloon_sprites[i].append(None)

monkey_sprite = Sprite(2.5 * cell_width, 7.5 * cell_height, "monkey.png", 148, 116)
arm_sprite = Sprite(0, 0, "monkey_arm_centered.png", 196, 196)
arm_sprite.set_center(monkey_sprite.x - 25, monkey_sprite.y - 5)
dart_sprite = Sprite(0, 0, "dart.png", 136, 136)

#	Juego
run = True
while run:
	for event in pg.event.get():
		if event.type == pg.QUIT:
			run = False

	#keys = pg.key.get_pressed()
	[mouse_x, mouse_y] = mouse.get_pos()

	#left_onpress = not left_click
	left_click = mouse.get_pressed(3)[0]
	if left_click and not flying:
		flying = True
		dart_velocity[0] = dart_speed * np.cos(mouse_angle * np.pi / 180.0)
		dart_velocity[1] = dart_speed * np.sin(mouse_angle * np.pi / 180.0)
		magnitude = np.sqrt(dart_velocity[0]**2 + dart_velocity[1]**2)
		dart_sprite.x = monkey_sprite.x + dart_velocity[0] / magnitude
		dart_sprite.y = monkey_sprite.y + dart_velocity[1] / magnitude
		dart_sprite.update()

	#left_onpress = left_onpress and left_click
	right_click = mouse.get_pressed(3)[2]
	#if right_click:
	#	flying = False

	#if (left_click or right_click) and search_state[0] == 2:
	#	search_state[0] = 0
	#	state = "Idle"
	#	state_color[0] = white
	#	state_text[0] = font.render(state, True, state_color[0], None)

	#check_button_hovers(mouse_x, mouse_y, left_click, left_onpress)

	mouse_angle = math.atan2(mouse_y - monkey_sprite.y, mouse_x - monkey_sprite.x) * 180.0 / np.pi

	if flying:
		arm_sprite.set_angle(mouse_angle)
	else:
		arm_sprite.set_angle(mouse_angle + 60)

	if not flying:
		arm_length = 30
		dart_sprite.x = arm_sprite.x + arm_length * np.cos(arm_sprite.angle * np.pi / 180.0)
		dart_sprite.y = arm_sprite.y + arm_length * np.sin(arm_sprite.angle * np.pi / 180.0)
		dart_sprite.update()
		dart_sprite.set_angle(mouse_angle)
	else:
		#dart_velocity[0] -= (dart_velocity[0] * dart_velocity[0]) * 0.0002 * delta_time	#roze con el aire en la componente horizontal de la velocidad
		dart_velocity[1] += dart_gravity * delta_time
		dart_velocity_angle = math.atan2(dart_velocity[1], dart_velocity[0]) * 180 / np.pi
		dart_sprite.x += dart_velocity[0] * delta_time
		dart_sprite.y += dart_velocity[1] * delta_time
		dart_sprite.update()
		dart_sprite.set_angle(dart_velocity_angle)

		tip = [0, 0]
		tip[0] = dart_sprite.x + dart_sprite.w / 2 * np.cos(dart_velocity_angle * np.pi / 180.0)
		tip[1] = dart_sprite.y + dart_sprite.w / 2 * np.sin(dart_velocity_angle * np.pi / 180.0)

		pop = [-1, -1]
		pop[0] = int(tip[0] / cell_width)
		pop[1] = int(tip[1] / cell_height)

		if pop[0] >= 0 and pop[0] < columns and pop[1] >= 0 and pop[1] < rows and bloons_data[pop[0]][pop[1]] == 1:
			bloons_data[pop[0]][pop[1]] = 0
			bloon_sprites[pop[0]][pop[1]] = None

	if dart_sprite.x > width or dart_sprite.y > height:
		flying = False

	draw()
	delta_time = clock.tick_busy_loop(framerate) / 1000.0
	pg.display.set_caption("Bloons AI - FPS: " + str(int(clock.get_fps())))

pg.quit()